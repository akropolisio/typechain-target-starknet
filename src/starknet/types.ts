import BN, { isBN } from 'bn.js';
import { AddTransactionResponse, Contract, InvokeFunctionTransaction } from 'starknet';

export interface ContractWrapper extends Omit<Contract, 'call' | 'invoke'> {
  call: (...args: any[]) => any;
  invoke: (...args: any[]) => any;
  getInvokeFunctionTransaction: (...args: any[]) => any;
}

type JSType = string | BN;
export function isPlainValue<T>(value: T): value is Extract<T, JSType> {
  return typeof value === 'string' || isBN(value);
}

const abiTypes = ['felt', 'Uint256'] as const;
export type AbiType = typeof abiTypes[number];
export function isAbiType(value: unknown): value is AbiType {
  return abiTypes.includes(value as any);
}

type AbiTypeMap<K, T extends Record<AbiType, K>> = Pick<T, AbiType>;
export type AbiToInputJSTypeMap = AbiTypeMap<
  JSType,
  {
    felt: string;
    Uint256: BN;
  }
>;

export type AbiToStarknetResponseTypeMap = AbiTypeMap<
  string[],
  {
    felt: [string];
    Uint256: [string, string];
  }
>;

export type EnvArgs = { [key in string]: JSType | EnvArgs };

export type CallFunction = (method: string, args?: EnvArgs) => Promise<EnvArgs | void>;
export type InvokeFunction = (method: string, args?: EnvArgs) => Promise<AddTransactionResponse>;
export type GetInvokeFunctionTransaction = (
  method: string,
  args?: EnvArgs,
) => InvokeFunctionTransaction;

export type StarknetCallData = string[];
