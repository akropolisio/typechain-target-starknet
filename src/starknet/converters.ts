import { Contract, AbiEntry, uint256, CallContractResponse, number } from 'starknet';

import {
  EnvArgs,
  isAbiType,
  StarknetCallData,
  AbiType,
  isPlainValue,
  AbiToInputJSTypeMap,
  AbiToStarknetResponseTypeMap,
} from './types';

export function inputsToCallData(
  inputsAbi: AbiEntry[],
  args: EnvArgs,
  structs: Contract['structs'],
): StarknetCallData {
  return inputsAbi.flatMap(({ name, type }) => {
    const arg = args[name];
    if (!arg) {
      throw new Error(`ABI parameter ${name} not found in ${args}`);
    }
    const struct = structs[type];
    if (!struct && !isAbiType(type)) {
      throw new Error(`Unknown abi type ${type}`);
    }
    return isAbiType(type)
      ? convertToCallData(arg, type)
      : inputsToCallData(struct.members, arg as any, structs);
  });
}

export function responseToOutput(
  outputsAbi: AbiEntry[],
  response: CallContractResponse,
  structs: Contract['structs'],
): EnvArgs {
  const { result } = outputsAbi.reduce(
    (acc, { name, type }) => {
      const struct = structs[type];
      const chunkLength = struct ? struct.members.length : 1;
      const nextLastArgIndex = acc.lastArgIndex + chunkLength;

      const arg = response.result.slice(acc.lastArgIndex, nextLastArgIndex);

      if (!arg?.length) {
        throw new Error(`ABI parameter ${name} not found in ${response}`);
      }
      if (!struct && !isAbiType(type)) {
        throw new Error(`Unknown abi type ${type}`);
      }

      return {
        lastArgIndex: nextLastArgIndex,
        result: {
          ...acc.result,
          [name]: isAbiType(type)
            ? convertToResponseValue(arg, type)
            : responseToOutput(struct.members, { result: arg }, structs),
        },
      };
    },
    { lastArgIndex: 0 } as { lastArgIndex: number; result: EnvArgs },
  );
  return result;
}

function convertToResponseValue(value: string | string[], abiType: AbiType): EnvArgs[string] {
  const toResponse: {
    [key in AbiType]: (v: AbiToStarknetResponseTypeMap[key]) => EnvArgs[string];
  } = {
    felt: v => v[0],
    Uint256: v => {
      const [low, high] = v;
      return uint256.uint256ToBN({ low, high });
    },
  };

  return (toResponse[abiType] as any)(value);
}

function convertToCallData(value: EnvArgs[string], abiType: AbiType): StarknetCallData {
  if (!isPlainValue(value)) {
    throw new Error(`Expected string, or BN, but received ${value}`);
  }
  const toInvokeCalldata: {
    [key in AbiType]: (input: AbiToInputJSTypeMap[key]) => StarknetCallData;
  } = {
    felt: v => {
      return [number.toBN(v).toString()];
    },
    Uint256: v => {
      const { low, high } = uint256.bnToUint256(v);
      return [number.toBN(low).toString(), number.toBN(high).toString()];
    },
  };
  return (toInvokeCalldata[abiType] as any)(value);
}
