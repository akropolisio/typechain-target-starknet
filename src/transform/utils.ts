import { getFilename, getUsedIdentifiers } from 'typechain';

export function getFilesParams(path: string) {
  const fullName = getFilename(path);
  const shortName = fullName.split('.')[0];
  const ShortName = `${shortName.slice(0, 1).toUpperCase()}${shortName.slice(1)}`;

  return {
    shortName,
    creatorName: `createStark${ShortName}`,
    contractType: `Stark${ShortName}Contract`,
    abiPath: `./abi/${fullName}`,
  };
}

type ModuleSpecifier = string;
type Name = string;
type IsDefault = true;
type Identifier = [Name, IsDefault] | Name;

export function createUsedImports(
  possibleImports: Record<ModuleSpecifier, Identifier[]>,
  sourceFile: string,
) {
  return Object.entries(possibleImports)
    .map(([moduleSpecifier, identifiers]) => {
      const usedIdentifiers = getUsedIdentifiers(
        identifiers.map(x => getIdentifierName(x)),
        sourceFile,
      );
      return createImportDeclaration(
        identifiers.filter(x => usedIdentifiers.includes(getIdentifierName(x))),
        moduleSpecifier,
      );
    })
    .filter(x => x)
    .join('\n');
}

function createImportDeclaration(identifiers: Identifier[], moduleSpecifier: string) {
  const [defaultImport] = identifiers.find(x => typeof x !== 'string') || [];
  const restIdentifiers = identifiers.filter(x => typeof x === 'string');

  const namedImports = restIdentifiers.length > 0 ? `{ ${restIdentifiers.join(', ')} }` : '';
  const allImports = [defaultImport, namedImports].filter(x => !!x).join(', ');
  return allImports.length ? `import ${allImports} from '${moduleSpecifier}'` : '';
}

function getIdentifierName(identifier: Identifier) {
  return typeof identifier === 'string' ? identifier : identifier[0];
}
