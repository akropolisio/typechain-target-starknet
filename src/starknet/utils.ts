import { AbiEntry, FunctionAbi } from 'starknet';

export function extractAbiMethod(abi: AbiEntry[], method: string) {
  const methodEntry = abi.find((m): m is FunctionAbi => m.name === method && m.type === 'function');
  if (!methodEntry) {
    throw new Error(`Method ${method} not found`);
  }
  return methodEntry;
}
