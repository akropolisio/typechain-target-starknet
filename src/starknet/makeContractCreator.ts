/* eslint-disable import/no-extraneous-dependencies */
import { Contract, Provider, stark } from 'starknet';

import { inputsToCallData, responseToOutput } from './converters';
import {
  CallFunction,
  ContractWrapper,
  EnvArgs,
  GetInvokeFunctionTransaction,
  InvokeFunction,
} from './types';
import { extractAbiMethod } from './utils';

export function makeContractCreator<T extends ContractWrapper>(abi: any[]) {
  return (address: string, provider: Provider): T => {
    const baseContract = new Contract(abi, address, provider);

    return (new Proxy<Contract>(baseContract, {
      get(target, prop: keyof ContractWrapper) {
        if (prop === 'call') {
          return makeCallFunction(baseContract, address);
        }
        if (prop === 'invoke') {
          return makeInvokeFunction(baseContract, address);
        }
        if (prop === 'getInvokeFunctionTransaction') {
          return makeGetInvokeFunctionTransaction(baseContract, address);
        }
        return target[prop];
      },
    }) as unknown) as T;
  };
}

function makeCallFunction(baseContract: Contract, contractAddress: string): CallFunction {
  return async (method: string, inputs?: EnvArgs) => {
    const methodAbi = extractAbiMethod(baseContract.abi, method);
    const requestInputs = inputs
      ? inputsToCallData(methodAbi.inputs, inputs, baseContract.structs)
      : [];

    const result = await baseContract.provider.callContract({
      contract_address: contractAddress,
      entry_point_selector: stark.getSelectorFromName(method),
      calldata: requestInputs,
    });
    const responseOutputs = responseToOutput(methodAbi.outputs, result, baseContract.structs);

    return responseOutputs;
  };
}

function makeInvokeFunction(baseContract: Contract, contractAddress: string): InvokeFunction {
  return (method: string, inputs?: EnvArgs) => {
    const methodAbi = extractAbiMethod(baseContract.abi, method);
    const requestInputs = inputs
      ? inputsToCallData(methodAbi.inputs, inputs, baseContract.structs)
      : [];

    return baseContract.provider.invokeFunction(
      contractAddress,
      stark.getSelectorFromName(method),
      requestInputs,
    );
  };
}

function makeGetInvokeFunctionTransaction(
  baseContract: Contract,
  contractAddress: string,
): GetInvokeFunctionTransaction {
  return (method: string, inputs?: EnvArgs) => {
    const methodAbi = extractAbiMethod(baseContract.abi, method);
    const requestInputs = inputs
      ? inputsToCallData(methodAbi.inputs, inputs, baseContract.structs)
      : [];

    return {
      type: 'INVOKE_FUNCTION',
      contract_address: contractAddress,
      entry_point_selector: stark.getSelectorFromName(method),
      calldata: requestInputs,
    };
  };
}
