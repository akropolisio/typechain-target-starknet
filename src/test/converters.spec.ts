import BN from 'bn.js';
import { StructAbi } from 'starknet';
import { encodeShortString } from 'starknet/dist/utils/shortString';

import { inputsToCallData, responseToOutput } from 'starknet/converters';
import { extractAbiMethod } from 'starknet/utils';

import testAbi from './abi/testAbi.json';

const structs = testAbi.reduce(
  (acc, curr) => (curr.type === 'struct' ? { ...acc, [curr.name]: curr as StructAbi } : acc),
  {} as Record<string, StructAbi>,
);

describe('inputsToCallData', () => {
  it('should correctly convert BN input to [low, high]', async () => {
    const args = { input_a: new BN('190000000000000000000') };
    const abi = extractAbiMethod(testAbi, 'test_b');
    const result = inputsToCallData(abi.inputs, args, structs);

    expect({ abi, args, result }).toMatchSnapshot();
  });

  it('should correctly convert multiple input values', async () => {
    const args = { input_a: encodeShortString('string'), input_b: new BN('190000000000000000000') };
    const abi = extractAbiMethod(testAbi, 'test_d');
    const result = inputsToCallData(abi.inputs, args, structs);

    expect({ abi, args, result }).toMatchSnapshot();
  });

  it('should correctly convert structures', async () => {
    const args = {
      input_a: {
        member_a: encodeShortString('string'),
        member_b: encodeShortString('string'),
        member_c: new BN('190000000000000000000'),
      },
    };
    const abi = extractAbiMethod(testAbi, 'test_a');
    const result = inputsToCallData(abi.inputs, args, structs);

    expect({ abi, args, result }).toMatchSnapshot();
  });
});

describe('responseToOutput', () => {
  it('should correctly convert [low, high] response to BN', async () => {
    const response = { result: ['0xa4cc799563c380000', '0x0'] };
    const abi = extractAbiMethod(testAbi, 'test_b');
    const result = responseToOutput(abi.outputs, response, structs);

    expect({ abi, response, result }).toMatchSnapshot();
  });

  it('should correctly convert multiple values in response', async () => {
    const response = {
      result: ['0xa4cc799563c380000', '0x0'],
    };
    const abi = extractAbiMethod(testAbi, 'test_d');
    const result = responseToOutput(abi.inputs, response, structs);

    expect({ abi, response, result }).toMatchSnapshot();
  });

  it('should correctly convert structures', async () => {
    const response = {
      result: [
        encodeShortString('string'),
        encodeShortString('string'),
        '0xa4cc799563c380000',
        '0x0',
      ],
    };
    const abi = extractAbiMethod(testAbi, 'test_a');
    const result = responseToOutput(abi.inputs, response, structs);

    expect({ abi, response, result }).toMatchSnapshot();
  });
});
