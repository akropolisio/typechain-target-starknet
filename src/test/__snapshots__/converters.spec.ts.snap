// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inputsToCallData should correctly convert BN input to [low, high] 1`] = `
Object {
  "abi": Object {
    "inputs": Array [
      Object {
        "name": "input_a",
        "type": "Uint256",
      },
    ],
    "name": "test_b",
    "outputs": Array [
      Object {
        "name": "output_a",
        "type": "Uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  "args": Object {
    "input_a": "0a4cc799563c380000",
  },
  "result": Array [
    "190000000000000000000",
    "0",
  ],
}
`;

exports[`inputsToCallData should correctly convert multiple input values 1`] = `
Object {
  "abi": Object {
    "inputs": Array [
      Object {
        "name": "input_a",
        "type": "felt",
      },
      Object {
        "name": "input_b",
        "type": "Uint256",
      },
    ],
    "name": "test_d",
    "outputs": Array [
      Object {
        "name": "output_a",
        "type": "felt",
      },
      Object {
        "name": "output_b",
        "type": "Uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  "args": Object {
    "input_a": "0x737472696e67",
    "input_b": "0a4cc799563c380000",
  },
  "result": Array [
    "126943972912743",
    "190000000000000000000",
    "0",
  ],
}
`;

exports[`inputsToCallData should correctly convert structures 1`] = `
Object {
  "abi": Object {
    "inputs": Array [
      Object {
        "name": "input_a",
        "type": "TestStruct",
      },
    ],
    "name": "test_a",
    "outputs": Array [
      Object {
        "name": "output_a",
        "type": "TestStruct",
      },
    ],
    "type": "function",
  },
  "args": Object {
    "input_a": Object {
      "member_a": "0x737472696e67",
      "member_b": "0x737472696e67",
      "member_c": "0a4cc799563c380000",
    },
  },
  "result": Array [
    "126943972912743",
    "126943972912743",
    "190000000000000000000",
    "0",
  ],
}
`;

exports[`responseToOutput should correctly convert [low, high] response to BN 1`] = `
Object {
  "abi": Object {
    "inputs": Array [
      Object {
        "name": "input_a",
        "type": "Uint256",
      },
    ],
    "name": "test_b",
    "outputs": Array [
      Object {
        "name": "output_a",
        "type": "Uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  "response": Object {
    "result": Array [
      "0xa4cc799563c380000",
      "0x0",
    ],
  },
  "result": Object {
    "output_a": "0a4cc799563c380000",
  },
}
`;

exports[`responseToOutput should correctly convert multiple values in response 1`] = `
Object {
  "abi": Object {
    "inputs": Array [
      Object {
        "name": "input_a",
        "type": "felt",
      },
      Object {
        "name": "input_b",
        "type": "Uint256",
      },
    ],
    "name": "test_d",
    "outputs": Array [
      Object {
        "name": "output_a",
        "type": "felt",
      },
      Object {
        "name": "output_b",
        "type": "Uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  "response": Object {
    "result": Array [
      "0xa4cc799563c380000",
      "0x0",
    ],
  },
  "result": Object {
    "input_a": "0xa4cc799563c380000",
    "input_b": "00",
  },
}
`;

exports[`responseToOutput should correctly convert structures 1`] = `
Object {
  "abi": Object {
    "inputs": Array [
      Object {
        "name": "input_a",
        "type": "TestStruct",
      },
    ],
    "name": "test_a",
    "outputs": Array [
      Object {
        "name": "output_a",
        "type": "TestStruct",
      },
    ],
    "type": "function",
  },
  "response": Object {
    "result": Array [
      "0x737472696e67",
      "0x737472696e67",
      "0xa4cc799563c380000",
      "0x0",
    ],
  },
  "result": Object {
    "input_a": Object {
      "member_a": "0x737472696e67",
      "member_b": "0x737472696e67",
      "member_c": "0a4cc799563c380000",
    },
  },
}
`;
